In [1]:
import numpy as np
import heapq
import math

In [2]:
# 1. Represent the grid as a graph using the 8-direction movement model 
GRID_SIZE = 201  # Grid from (0,0) to (200,200)
start_node = (0, 0)
goal_node = (200, 200)
# Movement directions and costs
dirs = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
costs = [1, 1, 1, 1, math.sqrt(2), math.sqrt(2), math.sqrt(2), math.sqrt(2)]

def is_obstacle(r, c):
    # Obstacle 1: (40, 40) to (70, 70)
    if 40 <= r <= 70 and 40 <= c <= 70: return True
    # Obstacle 2: (90, 120) to (130, 160)
    if 90 <= r <= 130 and 120 <= c <= 160: return True
    # Obstacle 3: (150, 50) to (180, 90)
    if 150 <= r <= 180 and 50 <= c <= 90: return True
    # Obstacle 4: (60, 150) to (100, 190)
    if 60 <= r <= 100 and 150 <= c <= 190: return True
    return False

In [3]:
# 2. Dijkstra's Algorithm Implementation
def dijkstra(start, goal):
    pq = [(0, start)]
    distances = {start: 0}
    expanded_nodes = 0
    visited = set()

    while pq:
        (dist, current) = heapq.heappop(pq)
        
        if current in visited: continue
        visited.add(current)
        expanded_nodes += 1
        
        if current == goal:
            return dist, expanded_nodes

        for (dr, dc), weight in zip(dirs, costs):
            nr, nc = current[0] + dr, current[1] + dc
            if 0 <= nr < GRID_SIZE and 0 <= nc < GRID_SIZE and not is_obstacle(nr, nc):
                new_dist = dist + weight
                if nr == 200 and nc == 200: # Final goal check
                    new_dist = dist + weight
                if (nr, nc) not in distances or new_dist < distances[(nr, nc)]:
                    distances[(nr, nc)] = new_dist
                    heapq.heappush(pq, (new_dist, (nr, nc)))
    return None, expanded_nodes

In [4]:
# 3. A* Algorithm Implementation
def diagonal_heuristic(node, goal):
    dx = abs(node[0] - goal[0])
    dy = abs(node[1] - goal[1])
    D = 1
    D2 = math.sqrt(2)
    return D * (dx + dy) + (D2 - 2 * D) * min(dx, dy)

def a_star(start, goal):
    pq = [(0 + diagonal_heuristic(start, goal), 0, start)]
    distances = {start: 0}
    expanded_nodes = 0
    visited = set()

    while pq:
        f, g, current = heapq.heappop(pq)
        
        if current in visited: continue
        visited.add(current)
        expanded_nodes += 1
        
        if current == goal:
            return g, expanded_nodes

        for (dr, dc), weight in zip(dirs, costs):
            nr, nc = current[0] + dr, current[1] + dc
            if 0 <= nr < GRID_SIZE and 0 <= nc < GRID_SIZE and not is_obstacle(nr, nc):
                new_g = g + weight
                if (nr, nc) not in distances or new_g < distances[(nr, nc)]:
                    distances[(nr, nc)] = new_g
                    f = new_g + diagonal_heuristic((nr, nc), goal)
                    heapq.heappush(pq, (f, new_g, (nr, nc)))
    return None, expanded_nodes

In [5]:
# Execution and Reporting
d_cost, d_expanded = dijkstra(start_node, goal_node)
a_cost, a_expanded = a_star(start_node, goal_node)

print(f"Dijkstra's Algorithm:")
print(f"Total Path Cost: {d_cost:.4f}") 
print(f"Number of Expanded Nodes: {d_expanded}") 

print(f"\nA* Algorithm:")
print(f"Total Path Cost: {a_cost:.4f}") #
print(f"Number of Expanded Nodes: {a_expanded}")

Dijkstra's Algorithm:
Total Path Cost: 301.0021
Number of Expanded Nodes: 34928

A* Algorithm:
Total Path Cost: 301.0021
Number of Expanded Nodes: 8134


The diagonal distance heuristic is admissible because it represents the actual shortest possible distance in an 8-direction grid without obstacles, thus it never overestimates the true remaining cost.

In [6]:
print(d_expanded - a_expanded)

26794


Comparison:

Path Length: Both algorithms found an identical shortest path cost of 301.0021.

Efficiency: A* expanded 26794 fewer nodes than Dijkstra, demonstrating higher efficiency through goal-directed search.